### ISSUE TITLE
Reduce and rationalize system library dependencies

### ISSUE THREAD

**Issue by pkgw** [tectonic-typesetting/tectonic#15]
We want Tectonic to depend on as few system libraries as we can get away with. This issue is for tracking the big picture of the overall project to reduce our dependence on system libraries.

The current roster of `pkg-config` dependencies is:

- fontconfig
- harfbuzz, harfbuzz-icu
- icu-uc
- freetype2
- graphite2
- libpng
- poppler
- zlib

Other libraries that the executable ends up depending on (on Linux):

- stdc++
- ssl, crypto
- xml2
- lzma
- iconv
- glib
- pcre
- tiff
- jpeg
- curl
- openjp2

We can probably eventually get rid of many of these!

**Comment by pkgw** [comment]
## Poppler

Usage includes:

- `pdfimage.cpp`: PDFDoc.h, Stream.h, Catalog.h
- `XeTeX_ext.c`: poppler-config.h

So it's really all about embedding PDFs as images within other PDFs. All `pdfimage.cpp` does is parse the document ("all") and get the bounds of a specified page, or count the pages.

**Comment by khaledhosny** [comment]
I think (not sure though) xdvipdfmx has code to extract the bounding box of PDF files (among other image files) AKA the extractbb mode.

**Comment by khaledhosny** [comment]
Indeed, it can read the bounding box and the number of pages:

```
$ extractbb -O -p5 Downloads/sile.pdf 
%%Title: Downloads/sile.pdf
%%Creator: extractbb 20160307
%%BoundingBox: 0 0 595 842
%%HiResBoundingBox: 0.000000 0.000000 595.280000 841.890000
%%PDFVersion: 1.5
%%Pages: 78
%%CreationDate: Sun Apr 23 16:15:38 2017
```

**Comment by pkgw** [comment]
And yes, it was just an hour's work to remove Poppler as a dependency! This is great, since Poppler is/was quite heavyweight.

**Comment by pkgw** [comment]
## libpng

Libpng is only used directly in `dpx-pngimage.c`, which of course is all about reading PNG files to get their bounding boxes and embed them in PDF output files.

Crates.io shows a popular PNG crate. Unclear if it would have all of the bells and whistles that are exploited in the current xdvipdfmx code. Strangely there isn't obviously a Rust `libpng-sys` crate ... you'd think there would be? Maybe PNGs are too simple for everyone.

Urgency: unclear. Libpng is popular, but its binary compatibility seems unreliable.

**Comment by chris-morgan** [comment]
Things like libpng or whatever equivalent you settle on can safely be optional, too—not everyone needs images in their documents.

**Comment by pkgw** [comment]
@chris-morgan Yup! The hope is that one day they will all be Cargo "features" that can be activated or deactivated as circumstances dictate. It will take a lot of infrastructure work to get to the point where that is possible, though.
